src/components/services/ActionQueueService.js

```diff
@@ -36,48 +43,72 @@ export const ActionQueueService = () => {
-          executing[programId] = true
-
-          const { program, state } = queue[programId]
-
-          if (program.mockMode) {
-            const { nextState } = await mockActionProgram(account, program, state)
-            dispatch(updateActionProgramState(nextState))
-            executing[programId] = false
-            return
-          }
+  useEffect(() => {
+    if (queue == null) return
 
-          const { nextState } = await executeActionProgram(executionContext, program, state).catch((error: Error) => {
-            console.warn(new Error('Action Program Exception: ' + error.message))
-            console.error(error)
-            return {
-              nextState: {
-                ...state,
-                effect: { type: 'done', error }
-              }
-            }
+    const { account, clientId } = executionContext
+
+    // Loop function
+    const execute = async () => {
+      // Programs that require immediate attention
+      const urgentProgramIds = Object.keys(queue).filter(programId => {
+        const { state } = queue[programId]
+        // Don't execute programs which are assigned to another client
+        if (state.clientId !== clientId) return false
+        // Don't execute programs which are currently executing
+        if (state.executing) return false
+        // Don't execute programs which have not reached their scheduled time
+        if (!state.effective && state.nextExecutionTime > Date.now()) return false
+        return true
+      })
+      // Act on urgent programs
+      const promises = urgentProgramIds.map(async programId => {
+        const { program, state } = queue[programId]
+
+        // Set program state to executing
+        await dispatch(
+          updateActionProgramState({
+            ...state,
+            executing: true
           })
+        )
 
+        if (program.mockMode) {
+          const { nextState } = await mockActionProgram(account, program, state)
           // Update program state
-          dispatch(updateActionProgramState(nextState))
+          dispatch(updateActionProgramState({ ...nextState, executing: false }))
+          return
+        }
 
-          // Unset program to running
-          executing[programId] = false
+        const { nextState } = await executeActionProgram(executionContext, program, state).catch((error: Error) => {
+          console.warn(new Error('Action Program Exception: ' + error.message))
+          console.error(error)
+          return {
+            nextState: {
+              ...state,
+              effect: { type: 'done', error },
+              effective: true,
+              nextExecutionTime: NaN
+            }
+          }
         })
 
+        // Update program state
+        await dispatch(updateActionProgramState({ ...nextState, executing: false }))
+      })
       await Promise.all(promises)
     }
-  }, [queue])
 
+    // Setup loop
+    const intervalId = setInterval(execute, EXECUTION_INTERVAL)
```

Comment:
`setInterval` is dangerous, because it doesn't account for how long the task takes to run. If each run takes 2 seconds, but you schedule the thing to run every second, you end up with infinite overlapping copies. You can use the `makePeriodicTask` from edge-core-js repo to handle this, since it is async-aware and able to work around this problem.

Reply:
Hmm, I wonder if this will be an issue considering that I'm filtering out tasks which are "executing".